Mathlib/SetTheory/ZFC/Rank.lean

```diff
@@ -0,0 +1,190 @@
+  · apply max_le
+    · exact (rank_lt_of_mem (mem_insert x y)).succ_le
+    · exact rank_mono (subset_iff.2 fun z => mem_insert_of_mem x)
+
+@[simp]
+theorem rank_singleton : rank {x} = succ (rank x) :=
+  rank_insert.trans (by simp)
+
+theorem rank_pair : rank {x, y} = max (succ (rank x)) (succ (rank y)) := by
+  simp
+
+@[simp]
+theorem rank_powerset : rank (powerset x) = succ (rank x) := by
+  apply le_antisymm
+  · simp_rw [rank_le_iff, mem_powerset, lt_succ_iff]
+    intro
+    exact rank_mono
+  · rw [succ_le_iff]
+    apply rank_lt_of_mem
+    simp
+
+/-- `PSet.rank` is equal to the `WellFounded.rank` over `∈`. -/
+theorem rank_eq_wf_rank : lift.{u + 1, u} (rank x) = mem_wf.rank x := by
+  induction' x using mem_wf.induction with x ih
+  rw [mem_wf.rank_eq]
+  simp_rw [← fun y : { y // y ∈ x } => ih y y.2]
+  apply (le_of_forall_lt _).antisymm (Ordinal.sup_le _) <;> intro h
+  · rw [lt_lift_iff]
+    rintro ⟨o, rfl, h⟩
+    simpa [Ordinal.lt_sup] using lt_rank_iff.1 h
+  · simpa using rank_lt_of_mem h.2
+
+end PSet
+
+namespace ZFSet
+
+variable {x y : ZFSet.{u}}
+
+/-- The ordinal rank of a ZFC set -/
+noncomputable def rank : ZFSet.{u} → Ordinal.{u} :=
+  Quotient.lift _ fun _ _ => PSet.rank_eq_of_equiv
+
+theorem rank_lt_of_mem : y ∈ x → rank y < rank x :=
+  Quotient.inductionOn₂ x y fun _ _ => PSet.rank_lt_of_mem
+
+theorem rank_le_iff {o : Ordinal} : rank x ≤ o ↔ (∀ y ∈ x, rank y < o) :=
+  ⟨fun h _ h' => (rank_lt_of_mem h').trans_le h,
+    Quotient.inductionOn x fun _ h =>
+      PSet.rank_le_iff.2 fun y h' => h ⟦y⟧ h'⟩
+
+theorem lt_rank_iff {o : Ordinal} : o < rank x ↔ ∃ y ∈ x, o ≤ rank y := by
+  rw [← not_iff_not, not_lt, rank_le_iff]
+  simp
+
+theorem rank_mono (h : x ⊆ y) : rank x ≤ rank y :=
+  rank_le_iff.2 fun _ h₁ => rank_lt_of_mem (h h₁)
+
+@[simp]
+theorem rank_empty : rank ∅ = 0 := PSet.rank_empty
+
+@[simp]
+theorem rank_insert : rank (insert x y) = max (succ (rank x)) (rank y) :=
+  Quotient.inductionOn₂ x y fun _ _ => PSet.rank_insert
+
+@[simp]
+theorem rank_singleton : rank {x} = succ (rank x) :=
+  rank_insert.trans (by simp)
+
+theorem rank_pair : rank {x, y} = max (succ (rank x)) (succ (rank y)) := by
+  simp
+
+@[simp]
+theorem rank_union : rank (x ∪ y) = max (rank x) (rank y) := by
+  apply le_antisymm
+  · simp_rw [rank_le_iff, mem_union, lt_max_iff]
+    intro
+    apply Or.imp <;> apply rank_lt_of_mem
+  · apply max_le <;> apply rank_mono <;> intro _ h <;> simp [h]
+
+@[simp]
+theorem rank_powerset : rank (powerset x) = succ (rank x) :=
+  Quotient.inductionOn x fun _ => PSet.rank_powerset
+
+/-- For the rank of `⋃₀ x`, we only have `rank (⋃₀ x) ≤ rank x ≤ rank (⋃₀ x) + 1`.
+    This inequality is splitted into `rank_sUnion_le` and `succ_rank_sUnion_ge`. -/
+theorem rank_sUnion_le : rank (⋃₀ x) ≤ rank x := by
+  simp_rw [rank_le_iff, mem_sUnion]
+  intro _ ⟨_, _, _⟩
+  trans <;> apply rank_lt_of_mem <;> assumption
+
+theorem succ_rank_sUnion_ge : rank x ≤ succ (rank (⋃₀ x)) := by
```

Comment:
Don't use `ge` either in lemma statements or lemma names as people will be looking for `le` instead
```suggestion
theorem le_succ_rank_sUnion : rank x ≤ succ (rank (⋃₀ x)) := by
```